File: src/RailworksForge.Core/VehicleTemplates.cs
namespace RailworksForge.Core;

public class VehicleTemplates
{
    private static string EngineTemplate => GetTextResource("Engine.xml");
    private static string WagonTemplate => GetTextResource("Wagon.xml");
    private static string TenderTemplate => GetTextResource("Tender.xml");

    private static string GetTextResource(string name)
    {
        var assembly = typeof(VehicleTemplates).Assembly;
        var resource = assembly.GetManifestResourceStream($"RailworksForge.Core.Resources.{name}");

        if (resource is null)
        {
            throw new Exception($"could not resource in assembly with name {name}");
        }

        using var reader = new StreamReader(resource);

        return reader.ReadToEnd();
    }

    public static string GetXml(VehicleType type)
    {
        return type switch
        {
            VehicleType.Engine => EngineTemplate,
            VehicleType.Wagon => WagonTemplate,
            VehicleType.Tender => TenderTemplate,
            _ => throw new Exception("Unknown vehicle type."),
        };
    }
}

File: src-rust/src-tauri/src/services/vehicle_templates.rs
use crate::models::vehicle::BlueprintType;

const ENGINE_TEMPLATE: &str = include_str!("templates/Engine.xml");
const WAGON_TEMPLATE: &str = include_str!("templates/Wagon.xml");
const TENDER_TEMPLATE: &str = include_str!("templates/Tender.xml");

pub fn get_template(blueprint_type: &BlueprintType) -> &'static str {
    match blueprint_type {
        BlueprintType::Engine => ENGINE_TEMPLATE,
        BlueprintType::Tender => TENDER_TEMPLATE,
        _ => WAGON_TEMPLATE,
    }
}
